Fixed data not showing

maleb1337 opened the pull request:
- Added coroutines, dagger, refactor the architecture of project and databinding feature.

maleb1337 committed:
- Fixed data not showing.
- Added coroutines, dagger, refactor the architecture of project and databinding feature.

danponce commented on app/build.gradle:
```diff
@@ -40,22 +47,36 @@ dependencies {
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:$lifecycleVersion"
+    implementation "androidx.lifecycle:lifecycle-extensions:$lifecycleVersion"
+
+    // ImageLoading
```
Great that you establish an order in gradle imports!

danponce commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+                    viewModel.Date =
+                        viewModel.parseDateFromCalendar(dayOfMonth, monthOfYear + 1, year)
+
+                    viewModel.loadCovid().observe(viewLifecycleOwner, Observer {
```
This has to be improved. A new observer is being created each time the btnDateSelection is clicked

danponce replied:
The viewmodel needs to have a way to load new data but not each time create a new livedata object

vtchkn replied:
Here the thing should be like this: 

OnDateSetListener {
    viewModel.fetch(date)
}

And below where you have an observer, subscribe to the changes and update your layout

maleb1337 replied:
Now this has improved. In CovidViewModel I added a transformations var with switchmap on it, that every time a fetch a date or in this case my new var "dateLiveData" changes its value, the method will call the API with the date.

danponce commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+        return date
+    }
+
+    fun parseDateFromCalendar(dayOfMonth: Int, monthOfYear: Int, year: Int): String {
```
In the future this can be moved to a DateUtils class and keep viewmodel cleaner

maleb1337 replied:
Agreed.

vtchkn replied:
Same here

danponce commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+    private lateinit var binding: CovidLayoutBinding
+
+    // Calendar variables
+    val c = Calendar.getInstance()
```
Isn't there a way to move this variables to another class? CovidFragment shouldn't know about calendar variables

maleb1337 replied:
I think that I'll need to do DateUtils to return this data. Like functions based in calendar that returns day, month and year by separately. I need that because is required to form a datepicker dialog. What do you think ?

vtchkn replied:
You definitely need remove this from fragment, you can consider moving the dialog creation away from the fragment too, using a Kotlin extensions, for example, take a look at this page: https://android.jlelse.eu/android-custom-alert-dialogs-kotlin-extension-functions-kotlin-higher-order-functions-life-682305c5322e

maleb1337 replied:
Indeed, I removed completely from the Fragment

danponce commented on app/src/main/res/layout/covid_layout.xml:
```diff
@@ -0,0 +1,89 @@
+
+    <data>
+
+        <variable
```
In the future you could have a Covid object variable inside your **CovidViewModel**. This way you only use the viewmodel as a variable inside this xml layout. Once these changes are made you need to notify to the binding object that the Covid object has changed everytime you make a new call to the API

maleb1337 replied:
I added the necessary variables to display data (to be exact 3 variables. date, confirmedCases, deaths) in the viewModel. I didn't add the Covid object.

matiasarancibia replied:
It would be nice create and use the object that Dan mentioned above, because in that way the code is cleaner

danponce commented on app/src/main/res/layout/covid_layout.xml:
```diff
@@ -0,0 +1,89 @@
+                android:orientation="vertical">
+
+                <TextView
+                    android:id="@+id/tvConfirmedCases"
```
As you can see, the only difference in these 2 TextViews is the **id** and the **text**. Then you can create a **style** that has all others fields and use it in both Textviews

maleb1337 replied:
True.

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+    val month = c.get(Calendar.MONTH)
+    val day = c.get(Calendar.DAY_OF_MONTH)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
If you don't do any job in a method, don't override it

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+                    viewModel.loadCovid().observe(viewLifecycleOwner, Observer {
+                        when (it.status) {
+                            Result.Status.SUCCESS -> {
+                                viewModel.isShowProgress.value = View.GONE
```
I personally don't like the Int implementation in logically Boolean fields, as it then became very difficult to understand what's going on, so I would do this viewModel.isShowProgress = false
and if you like, you can create an extension, or do this job inside a layout with Data Binding

like this "@{viewModel.isShowProgress  == true ? View.VISIBLE : View.GONE}"

maleb1337 replied:
Now those variables are in Boolean and I added in CovidViewModel a method that determines their visibility within true or false.

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+                            Result.Status.SUCCESS -> {
+                                viewModel.isShowProgress.value = View.GONE
+                                viewModel.isShowContent.value = View.VISIBLE
+                                binding.covid = it.data
```
Also if you are using Data binding, it is not necessary to pass the data like this, you can receive it from viewModel, please take a look at the google Codelabs 
https://codelabs.developers.google.com/codelabs/android-databinding/index.html?index=..%2F..index#6

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+                            }
+                            Result.Status.LOADING -> {
+                                viewModel.isShowProgress.value = View.VISIBLE
+                                viewModel.isShowContent.value = View.GONE
```
isShowContent is not necessary, if it is not error you show content. Because you could have content showing **and** a progress bar

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+                                binding.covid = it.data
+                            }
+                            Result.Status.ERROR -> {
+                                viewModel.isShowProgress.value = View.GONE
```
If you have an error, you should work it. Maybe a toast or snackbar

maleb1337 replied:
I added a snackbar now, when it comes to errors.

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+
+    }
+
+    override fun onCreateView(
```
All this code I would pass to onViewCreated

maleb1337 replied:
Do you mean ALL of the onCreateView code?

vtchkn replied:
@maleb1337 yes, I can't see any specific task you may need to do _before_ creating the View

maleb1337 replied:
Yup. Now I removed onCreateView, now only I have onViewCreated with some adaptions regarding the layout binding.

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,112 @@
+        viewModel.covid.observe(viewLifecycleOwner, Observer {
+            when (it.status) {
+                Result.Status.SUCCESS -> {
+                    viewModel.isShowProgress.value = View.GONE
```
Same here as all above!

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+
+    fun getCurrentDay(): String {
+        val c = Calendar.getInstance()
+        val year = c.get(Calendar.YEAR)
```
I would create an extension and not doing this work here

Also, please, use DateFormatter! No hardcoded string is allowed

maleb1337 committed:
- Fixed some code based on comments.
- Having issues when it comes to "rebinding data" also with datepicker dialog...

maleb1337 committed:
- Fixed datepicker dialog problem.
- Updated CovidViewModel.

danponce reviewed (changes requested):
Pull request has to be more improved. Comments has to be answered and apply changes needed

maleb1337 committed:
- Only one observer: Now in CovidFragment there's only one observer and added to the viewModel a transformations var with switchMap (I did that based on actual CP code)

maleb1337 committed:
- Removed hardcoded strings.

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/api/AuthInterceptor.kt:
```diff
@@ -0,0 +1,16 @@
+    @Throws(IOException::class)
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val request = chain.request().newBuilder()
+            .addHeader("X-RapidAPI-KEY", "96afa298cbmsh913f910f914494cp110c39jsn01a32d68445e")
```
Would be great to use a key from Gradle.properties

maleb1337 replied:
Ok, I implemented a field in gradle.properties. Then added the app build.gradle the following buildConfigField("String", "API_KEY", API_KEY) and lastly when is time to use this simply gets called by Buildconfig.API_KEY.

Is that what you expected or am I wrong ?

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/data/CovidRepository.kt:
```diff
@@ -0,0 +1,33 @@
+            }
+            when (result.status) {
+                Result.Status.SUCCESS -> {
+                    emit(Result.success(result.data!!.data!!))
```
don't use !!, better do a null-safe ?, and if the content is null, emit the Result.Empty state

maleb1337 replied:
Added empty state and fixed from !! to ? null-safe

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/data/CovidRepository.kt:
```diff
@@ -0,0 +1,33 @@
+                    emit(Result.success(result.data!!.data!!))
+                }
+                Result.Status.ERROR -> {
+                    emit(Result.error(result.data!!.message!!))
```
Same here, you cannot be sure that in case of an error you will get any data, OR that this data will contain a message

maleb1337 replied:
Fixed now!

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,102 @@
+            .observe(viewLifecycleOwner, Observer {
+                when (it.status) {
+                    Result.Status.SUCCESS -> {
+                        viewModel.isShowProgress.value = false
```
you can move all the work you call in ViewModel inside the viewModel where you emit states

maleb1337 replied:
Ok !

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidViewModel.kt:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    // Every time that dateLiveData has another value, it calls the API.
```
Review what date is actually calling

maleb1337 replied:
I didn't understand this comment...

matiasarancibia replied:
Was this solved?

vtchkn commented on app/src/main/java/cl/maleb/testcarsales/utils/DateUtils.kt:
```diff
@@ -0,0 +1,86 @@
+    return date
+}
+
+fun datePickerDialog(context: Context, listener: DatePickerDialog.OnDateSetListener) {
```
I would change the name of the method to showDatePickerDialog

maleb1337 replied:
Okay !

maleb1337 committed:
- Added field in gradle.properties with API_KEY and complemented with build.gradle to use it
- Deleted onCreateView on CovidFragment
- Refactor name of fun datePickerDialog to showDatePickerDialog in DateUtils

maleb1337 committed:
- Minor fix in Fragment

maleb1337 committed:
- Added empty state.
- State value handler in ViewModel.

maleb1337 committed:
- Minor change in errorState

matiasarancibia commented on app/src/main/java/cl/maleb/testcarsales/data/Dummy.kt:
```diff
@@ -0,0 +1,3 @@
+package cl.maleb.testcarsales.data
+
+data class Dummy(val name: String)
```
What is the use for this class? Maybe you could change the name to be more specific and descriptive

maleb1337 replied:
The use of this class is for a di class called **CoreDataModule** I attached an image. And you're right I should've put a description.
![Screenshot_1](https://user-images.githubusercontent.com/23465142/89339916-eed57500-d66c-11ea-9be8-ae42cccccc9d.png)

matiasarancibia commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidFragment.kt:
```diff
@@ -0,0 +1,86 @@
+            .observe(viewLifecycleOwner, Observer {
+                when (it.status) {
+                    Result.Status.SUCCESS -> {
+                        viewModel.successState(it.data?.date, it.data?.confirmed, it.data?.deaths)
```
You could send only the value of `it.data` and handle each needed field inside the view model

maleb1337 replied:
I created a MutableLiveData<Covid>() variable to handle all the it.data in just one variable.

matiasarancibia commented on app/src/main/java/cl/maleb/testcarsales/ui/covid/CovidViewModel.kt:
```diff
@@ -0,0 +1,64 @@
+        return when (boolean) {
+            true -> View.VISIBLE
+            false -> View.GONE
+        }
```
All the logic related with the UI (items visibility, test size, set texts, etc) should be in the fragment and not in the view model

matiasarancibia commented on app/src/main/java/cl/maleb/testcarsales/utils/DateUtils.kt:
```diff
@@ -0,0 +1,86 @@
+    var dateString = ""
+
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val current = LocalDateTime.now().minusDays(1)
+        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+        dateString = current.format(formatter)
+    } else {
+        val date = Date(currentYear, currentMonth, currentDay)
+        val formatter = SimpleDateFormat("yyyy-MM-dd")
+        dateString = formatter.format(date)
+    }
+
+    return dateString
+}
+
+fun getCurrentDay(): String {
+    var dateString = ""
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val current = LocalDateTime.now()
+        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+        dateString = current.format(formatter)
+    } else {
+        val date = Date()
+        val formatter = SimpleDateFormat("yyyy-MM-dd")
+        dateString = formatter.format(date)
+    }
```
If this methods do the same thing, you can create a general method to handle the dates and use it wherever you need

matiasarancibia replied:
One option is convert this method to an extension of Context, in this way you can use that context to call the method `getString` and use a string resource instead of the hardcoded text

maleb1337 replied:
Yeah, I replicated the method and forgot to erase the one that I don't even use it

matiasarancibia commented on app/src/main/java/cl/maleb/testcarsales/utils/DateUtils.kt:
```diff
@@ -0,0 +1,86 @@
+    var date = ""
+    if (dayOfMonth < 10) {
+        if (monthOfYear < 10) {
+            date = "$year-0$monthOfYear-0$dayOfMonth"
+        } else {
+            date = "$year-$monthOfYear-0$dayOfMonth"
+        }
+
+    } else {
+        if (monthOfYear < 10) {
+            date = "$year-0$monthOfYear-$dayOfMonth"
+        } else {
+            date = "$year-$monthOfYear-$dayOfMonth"
```
Please use string resources instead of hardcoded texts

maleb1337 replied:
Whoops, it's fixed now

matiasarancibia commented on app/src/main/res/layout/covid_layout.xml:
```diff
@@ -0,0 +1,75 @@
+            <TextView
+                android:id="@+id/tvDate"
+                style="@style/textHeader"
+                android:text="@{@string/date + ' ' + viewModel.selectedDate}" />
```
Just like a suggestion... avoid of concatenate strings directly in the xml, I think it's a better idea use a binding adapter or format the text, save it into a variable and then just use it in the xml

maleb1337 committed:
- Optimization of code, that gets the data of the API.
- Cleaned DateUtils code and finally no more hardcoded strings.

maleb1337 committed:
- Minor changes and improvements regarding layout binding and showDatePickerDialog

maleb1337 committed:
- Code cleanup pt 2

maleb1337 committed:
- Prevent data loss when user rotates phone.
- Added scrollview for a better view in landscape.
- Calendar default date is minus one day.

maleb1337 committed:
- When the user has no internet connection. It displays a layout with the correct information.

maleb1337 committed:
- Now you can see the date field in landscape on covid_layout.xml.

maleb1337 committed:
- Number fields formatted in thousands.
- When the user selects a date greater o equal to the current date throws a warning message to the user instead of reloading

maleb1337 committed:
- Number format now is in dots instead of commas.
